Store/COSQLiteStorePersistentRootBackingStore.m

```diff
@@ -643,6 +659,106 @@ - (BOOL)writeItemGraph: (COItemGraph *)anItemTree
 }
 
+- (int64_t)rewriteRevid: (int64_t)revid
+          withItemGraph: (COItemGraph *)newItemGraph
+                version: (int64_t)newVersion
+     migratedItemGraphs: (NSDictionary *)migratedItemGraphs
+    lastBytesInDeltaRun: (int64_t)lastBytesInDeltaRun
+{
+    FMResultSet *rs = [db_ executeQuery:
+        [NSString stringWithFormat: @"SELECT parent, deltabase FROM %@ WHERE revid = ?", [self tableName]],
+        @(revid)];
+    int64_t bytesInDeltaRun = -1;
+
+    while ([rs next])
+    {
+        const int64_t parent = [rs longLongIntForColumnIndex: 0];
+        const int64_t deltabase = [rs longLongIntForColumnIndex: 1];
+        const int64_t parentDeltabase = [self deltabaseForRowid: parent];
+        const BOOL delta = parentDeltabase != -1 && deltabase == parentDeltabase;
+        NSData *contentsBlob;
+        
+        if (delta)
+        {
+            COItemGraph *parentItemGraph = migratedItemGraphs[@(parent)];
+            NSMutableSet *partialItems = [NSMutableSet setWithArray: newItemGraph.items];
+
+            // Remove all parent revision items that remain the same in the current revision
+            [partialItems minusSet: [NSSet setWithArray: parentItemGraph.items]];
+
+            COItemGraph *partialItemGraph =
+                [[COItemGraph alloc] initWithItems: partialItems.allObjects
+                                      rootItemUUID: newItemGraph.rootItemUUID];
+            
+            contentsBlob = contentsBLOBWithItemTree(partialItemGraph);
+            bytesInDeltaRun = lastBytesInDeltaRun + contentsBlob.length;
+        }
+        else
+        {
+            contentsBlob = contentsBLOBWithItemTree(newItemGraph);
+            bytesInDeltaRun = contentsBlob.length;
+        }
+        
+        BOOL ok = [db_ executeUpdate: [NSString stringWithFormat:
+            @"UPDATE %@ SET contents = ?, hash = ?, bytesInDeltaRun = ?, version = ? WHERE revid = ?",
+            [self tableName]],
+            contentsBlob, Sha1Data(contentsBlob), @(bytesInDeltaRun), @(newVersion), @(revid)];
+        
+        if (!ok)
+        {
+            [self rollback];
+            return -1;
+        }
+    }
+    
+    return bytesInDeltaRun;
+}
+
+- (int64_t)schemaVersionForRevid: (int64_t)revid
+{
+    NSNumber *version = [db_ numberForQuery:
+        [NSString stringWithFormat: @"SELECT version FROM %@ WHERE revid = ?", [self tableName]],
+        @(revid)];
+    if (version == nil)
+    {
+        return 0;
+    }
+    return version.unsignedLongLongValue;
+}
+
+- (BOOL)migrateRevisionsToVersion: (int64_t)newVersion withHandler: (COMigrationHandler)handler
+{
+    NSIndexSet *revids = [self revidsUsedRange];
+    NSMutableDictionary *migratedItemGraphs = [NSMutableDictionary dictionary];
+    int64_t __block lastBytesInDeltaRun = 0;
+
+    [self beginTransaction];
+
+    [revids enumerateIndexesUsingBlock: ^(NSUInteger revid, BOOL * _Nonnull stop) {
+        COItemGraph *oldItemGraph = [self itemGraphForRevid: revid];
```

Comment:
I think there's a bit of a problem here because `rewriteRevid:` doesn't implement the full rewrite logic that we were discussing (where child delta commits are fixed so they can return properly return the "old" commit contents).

Say you are migrating all 3 commits in the store, A (full snapshot), B (delta), C (delta).

This block runs for A, B, C, in that order.

Everything works fine for revision A, and it gets rewritten.

But when the block runs for B, this `-itemGraphForRevid: B` call is going to return a mix of the already-migrated items from A  plus some of the unmigrated items from B. Won't that be a problem for the `handler()` function?


I think if you want to go with the present approach here you'll need to use `-partialItemGraphFromRevid:toRevid:` to get *just* the items in the delta commit, that way you're guaranteed not to have a mix of item schema versions unlike with `-itemGraphForRevid:`.

With the approach in this PR it also means that `handler()` may be called with an incomplete item graph, so I think you can only support the "simple" type of migrations that are 1 item in -> the same 1 item out.

Reply:
> But when the block runs for B, this -itemGraphForRevid: B call is going to return a mix of the already-migrated items from A plus some of the unmigrated items from B. Won't that be a problem for the handler() function?

Thanks for pointing this out! I wrongly assumed that inside the transaction, queries would read the DB in the state prior to the transaction, but I checked SQLite documentation and that's not true: _Within a single database connection X, a SELECT statement always sees all changes to the database that are completed prior to the start of the SELECT statement, whether committed or uncommitted._

It's strange my migration test doesn't fail due to this, I'll need to take a better look.

To fix this, I need to either:

1. use the rewrite logic you suggested initially
2. rewrite revisions backwards (from most recent revision to the oldest one)

Option 1 is probably a bit slower because rewriting affects two revisions and each revision is rewritten twice. Option 2 is more complex to implement, because you need to pick a branch and a head revision to start with, then you have to check at every revision whether there is a child branch that needs to be rewritten first, before rewriting the previous revision on the branch you're processing.

I guess that's why you suggested option 1 initially.

Reply:
Yeah, I suggested option 1 because it keeps the store in a consistent state. You’re right it will rewrite each revision twice though so it is potentially a lot slower.

I think you can implement option 2 with a tiny change to your current code, just iterate the revids backwards (highest to lowest).